src/pool/mempool.rs

```diff
@@ -143,22 +144,37 @@ impl<SP: starknet::providers::Provider + Send + Sync + Clone + 'static> AccountM
     where
         SP: starknet::providers::Provider + Send + Sync + Clone + 'static,
     {
-        let mut accounts = self.accounts.iter().collect::<HashMap<_, _>>();
+        // Use `StdRng` instead of `ThreadRng` as it is `Send`
+        let mut rng = rand::rngs::StdRng::from_entropy();
+
+        // Collect the accounts into a vector for shuffling
+        let mut accounts: Vec<_> = self.accounts.iter().collect();
+
+        // Shuffle the accounts randomly before iterating
+        accounts.shuffle(&mut rng);
+
         loop {
             if accounts.is_empty() {
                 return Err(eyre::eyre!("failed to fetch funded account"));
             }
+
+            // Create the future locks with indices for more efficient removal
             // use [`select_all`] to poll an iterator over impl Future<Output = (Felt, MutexGuard<Felt>)>
             // We use Box::pin because this Future doesn't implement `Unpin`.
-            let fut_locks = accounts.iter().map(|(address, nonce)| Box::pin(async { (*address, nonce.lock().await) }));
-            let ((account_address, guard), _, _) = select_all(fut_locks).await;
+            let fut_locks = accounts
+                .iter()
+                .enumerate()
+                .map(|(index, (address, nonce))| Box::pin(async move { (index, *address, nonce.lock().await) }));
+
+            // Select the first account that gets unlocked
+            let ((index, account_address, guard), _, _) = select_all(fut_locks).await;
 
             // Fetch the balance of the selected account
             let balance = self.get_balance(*account_address).await?;
 
-            // If the balance is lower than the threshold, continue
+            // If the balance is lower than the threshold, remove the account using swap_remove
```

Comment:
If the balance is lower than the threshold, should we also alert? This should never happen ideally

Reply:
I think it's ok even if the balance goes below, the service will just go and get another relayer and it's ok because @eugypalu is building a rebalancing service which will allow to overcome this case even in the worst case to rebalance not long after if ever a relayer had to be emptied.

Reply:
@Eikix ok to merge?